File: CRM.Core/Models/Quote/Quote.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace CRM.Core.Models.Quote
{
    /// <summary>
    /// 报价单主表 (Quote)
    /// 对应数据库表: quote
    /// 说明: 一个报价单对应一个 RFQ 明细行（rfqitem），包含多个供应商报价行（quoteitem）
    /// </summary>
    [Table("quote")]
    public class Quote : BaseGuidEntity
    {
        [Key]
        [StringLength(36)]
        [Column("QuoteId")]
        public override string Id { get; set; } = Guid.NewGuid().ToString();

        /// <summary>报价单号，如 QT0001</summary>
        [Required]
        [StringLength(32)]
        [Column("quote_code")]
        public string QuoteCode { get; set; } = string.Empty;

        /// <summary>关联 RFQ 主表 ID</summary>
        [StringLength(36)]
        [Column("rfq_id")]
        public string? RFQId { get; set; }

        /// <summary>关联 RFQ 明细行 ID（一个报价单对应一个 rfqitem）</summary>
        [StringLength(36)]
        [Column("rfq_item_id")]
        public string? RFQItemId { get; set; }

        /// <summary>物料型号（冗余自 rfqitem，方便查询）</summary>
        [StringLength(200)]
        [Column("mpn")]
        public string? Mpn { get; set; }

        /// <summary>客户 ID</summary>
        [StringLength(36)]
        [Column("customer_id")]
        public string? CustomerId { get; set; }

        /// <summary>业务员 ID</summary>
        [StringLength(36)]
        [Column("sales_user_id")]
        public string? SalesUserId { get; set; }

        /// <summary>采购员 ID</summary>
        [StringLength(36)]
        [Column("purchase_user_id")]
        public string? PurchaseUserId { get; set; }

        /// <summary>报价日期</summary>
        [Column("quote_date")]
        public DateTime QuoteDate { get; set; } = DateTime.UtcNow;

        /// <summary>状态 (0:草稿 1:待审核 2:已审核 3:已发送 4:已接受 5:已拒绝 6:已过期 7:已关闭)</summary>
        [Column("status")]
        public short Status { get; set; } = 0;

        /// <summary>备注</summary>
        [StringLength(1000)]
        [Column("remark")]
        public string? Remark { get; set; }

        // 导航属性
        public virtual ICollection<QuoteItem> Items { get; set; } = new List<QuoteItem>();
    }

    /// <summary>
    /// 报价明细行 (QuoteItem)
    /// 对应数据库表: quoteitem
    /// 说明: 每行对应一个供应商的报价，包含价格、库存、交期等信息
    /// </summary>
    [Table("quoteitem")]
    public class QuoteItem : BaseGuidEntity
    {
        [Key]
        [StringLength(36)]
        [Column("QuoteItemId")]
        public override string Id { get; set; } = Guid.NewGuid().ToString();

        /// <summary>所属报价单 ID（外键）</summary>
        [Required]
        [StringLength(36)]
        [Column("quote_id")]
        public string QuoteId { get; set; } = string.Empty;

        // ─── 供应商信息 ───────────────────────────────────────────────

        /// <summary>供应商 ID</summary>
        [StringLength(36)]
        [Column("vendor_id")]
        public string? VendorId { get; set; }

        /// <summary>供应商名称（如 Digikey）</summary>
        [StringLength(200)]
        [Column("vendor_name")]
        public string? VendorName { get; set; }

        /// <summary>供应商代码（如 VH0LUA）</summary>
        [StringLength(50)]
        [Column("vendor_code")]
        public string? VendorCode { get; set; }

        /// <summary>联系人 ID</summary>
        [StringLength(36)]
        [Column("contact_id")]
        public string? ContactId { get; set; }

        /// <summary>联系人名称（如 Internatinal Sales）</summary>
        [StringLength(100)]
        [Column("contact_name")]
        public string? ContactName { get; set; }

        // ─── 价格类型 ─────────────────────────────────────────────────

        /// <summary>价格类型（如：现货价、期货价、样品价等）</summary>
        [StringLength(50)]
        [Column("price_type")]
        public string? PriceType { get; set; }

        /// <summary>失效日期</summary>
        [Column("expiry_date")]
        public DateTime? ExpiryDate { get; set; }

        // ─── 物料信息 ─────────────────────────────────────────────────

        /// <summary>物料型号（MPN，如 REF3430QDBVRQ1）</summary>
        [StringLength(200)]
        [Column("mpn")]
        public string? Mpn { get; set; }

        /// <summary>品牌（如 TEXAS INSTRUMENTS/德州仪器）</summary>
        [StringLength(200)]
        [Column("brand")]
        public string? Brand { get; set; }

        /// <summary>品牌属地（如 美国）</summary>
        [StringLength(100)]
        [Column("brand_origin")]
        public string? BrandOrigin { get; set; }

        // ─── 时效信息 ─────────────────────────────────────────────────

        /// <summary>生产日期（如 NA、2年内）</summary>
        [StringLength(100)]
        [Column("date_code")]
        public string? DateCode { get; set; }

        /// <summary>交期（如 10-12days DC下单限制）</summary>
        [StringLength(200)]
        [Column("lead_time")]
        public string? LeadTime { get; set; }

        // ─── 涂标 / 产地 ──────────────────────────────────────────────

        /// <summary>涂标 (0:不涂标 1:涂标 2:待确定)</summary>
        [Column("label_type")]
        public short LabelType { get; set; } = 2;

        /// <summary>报价晶圆产地 (0:美产 1:非美产 2:待确定)</summary>
        [Column("wafer_origin")]
        public short WaferOrigin { get; set; } = 2;

        /// <summary>报价封装产地 (0:美产 1:非美产 2:待确定)</summary>
        [Column("package_origin")]
        public short PackageOrigin { get; set; } = 2;

        /// <summary>是否包邮</summary>
        [Column("free_shipping")]
        public bool FreeShipping { get; set; } = false;

        // ─── 价格信息 ─────────────────────────────────────────────────

        /// <summary>报价币别 (统一币别编码：1=RMB 2=USD 3=EUR 4=HKD 5=JPY 6=GBP)</summary>
        [Column("currency")]
        public short Currency { get; set; } = 1;

        /// <summary>报价数量</summary>
        [Column("quantity", TypeName = "numeric(18,4)")]
        public decimal Quantity { get; set; } = 0;

        /// <summary>报价单价（含6位小数，如 2.267910）</summary>
        [Column("unit_price", TypeName = "numeric(18,6)")]
        public decimal UnitPrice { get; set; } = 0;

        /// <summary>折算价（按汇率折算后的价格）</summary>
        [Column("converted_price", TypeName = "numeric(18,6)")]
        public decimal? ConvertedPrice { get; set; }

        // ─── 库存 / 订购信息 ──────────────────────────────────────────

        /// <summary>最小包装数量</summary>
        [Column("min_package_qty")]
        public int MinPackageQty { get; set; } = 0;

        /// <summary>最小包装单位（如 Reel、Tube、Tray）</summary>
        [StringLength(50)]
        [Column("min_package_unit")]
        public string? MinPackageUnit { get; set; }

        /// <summary>库存数量</summary>
        [Column("stock_qty")]
        public int StockQty { get; set; } = 0;

        /// <summary>起订量（MOQ）</summary>
        [Column("moq")]
        public int Moq { get; set; } = 0;

        /// <summary>备注（如 散料）</summary>
        [StringLength(500)]
        [Column("remark")]
        public string? Remark { get; set; }

        /// <summary>状态 (0:有效 1:已取消)</summary>
        [Column("status")]
        public short Status { get; set; } = 0;

        // 导航属性
        public virtual Quote? Quote { get; set; }
    }
}

File: CRM.Web/src/api/quote.ts
import apiClient from './client'

const getYYMMDD = (d: Date) => {
  const yy = String(d.getFullYear()).slice(-2)
  const mm = String(d.getMonth() + 1).padStart(2, '0')
  const dd = String(d.getDate()).padStart(2, '0')
  return `${yy}${mm}${dd}`
}

const random4 = () => String(Math.floor(Math.random() * 10000)).padStart(4, '0')

/** 与后端 CreateQuoteItemRequest / 前端表单字段对齐 */
function buildItemsFromQuoteCreateForm(form: Record<string, unknown>): Record<string, unknown>[] {
  const rows = (form.quotePriceRows as Record<string, unknown>[] | undefined) || []
  const items: Record<string, unknown>[] = []
  for (const r of rows) {
    const qty = Number(r.quantity)
    const up = r.unitPrice != null && r.unitPrice !== '' ? Number(r.unitPrice) : NaN
    if (qty < 1 || Number.isNaN(up)) continue
    items.push({
      vendorId: form.vendorId || undefined,
      vendorName: (form.vendorName as string) || undefined,
      vendorCode: undefined,
      contactId: form.vendorContactId || undefined,
      contactName: (form.contactName as string) || undefined,
      priceType: (form.priceType as string) || undefined,
      expiryDate: form.expiryDate ? new Date(String(form.expiryDate)).toISOString() : undefined,
      mpn: (form.mpn as string) || undefined,
      brand: (form.brand as string) || undefined,
      brandOrigin: (form.brandOrigin as string) || undefined,
      dateCode: (form.productionDate as string) || undefined,
      leadTime: (form.leadTime as string) || undefined,
      labelType: Number(form.labelType ?? 0),
      waferOrigin: Number(form.waferOrigin ?? 2),
      packageOrigin: Number(form.packageOrigin ?? 2),
      freeShipping: Boolean(form.freeShipping),
      currency: Number(r.currency ?? 1),
      quantity: qty,
      unitPrice: up,
      convertedPrice:
        r.convertedPrice != null && r.convertedPrice !== '' ? Number(r.convertedPrice) : undefined,
      minPackageQty: Number(form.minPackageQty ?? 0),
      minPackageUnit: undefined,
      stockQty: Number(form.stockQty ?? 0),
      moq: Number(form.moq ?? 0),
      remark: (form.remark as string) || undefined,
      status: 0
    })
  }
  return items
}

function mapToCreateQuoteRequest(form: Record<string, unknown>): Record<string, unknown> {
  const items = buildItemsFromQuoteCreateForm(form)
  if (items.length === 0) {
    throw new Error('至少需要一条有效的供应商报价明细（数量≥1 且单价有效）')
  }

  const quoteCode =
    (form.quoteCode && String(form.quoteCode).trim()) ||
    `QT${getYYMMDD(new Date())}${random4()}`

  return {
    quoteCode,
    rfqId: form.rfqId || null,
    rfqItemId: form.rfqItemId || null,
    mpn: (form.mpn as string) || null,
    customerId: form.customerId || null,
    salesUserId: form.salesUserId || null,
    purchaseUserId: form.purchaseUserId || null,
    quoteDate: (form.quoteDate as string) || new Date().toISOString().slice(0, 10),
    status: 0,
    remark: (form.remark as string) || null,
    items
  }
}

function mapQuoteEditSimpleToUpdate(body: Record<string, unknown>): Record<string, unknown> {
  const rawItems = (body.items as Record<string, unknown>[]) || []
  const items = rawItems.map((it) => ({
    vendorId: it.vendorId,
    vendorName: it.vendorName,
    vendorCode: it.vendorCode,
    contactId: it.contactId,
    contactName: it.contactName,
    priceType: it.priceType,
    mpn: it.mpn ?? body.mpn,
    brand: it.brand,
    brandOrigin: it.brandOrigin,
    dateCode: it.dateCode,
    leadTime: it.leadTime,
    labelType: Number(it.labelType ?? 2),
    waferOrigin: Number(it.waferOrigin ?? 2),
    packageOrigin: Number(it.packageOrigin ?? 2),
    freeShipping: Boolean(it.freeShipping ?? false),
    currency: Number(it.currency ?? 1),
    quantity: Number(it.quantity),
    unitPrice: Number(it.unitPrice),
    convertedPrice: it.convertedPrice != null ? Number(it.convertedPrice) : undefined,
    minPackageQty: Number(it.minPackageQty ?? 0),
    minPackageUnit: it.minPackageUnit,
    stockQty: Number(it.stockQty ?? 0),
    moq: Number(it.moq ?? 0),
    remark: it.remark,
    status: Number(it.status ?? 0)
  }))
  return {
    mpn: body.mpn,
    remark: body.remark ?? null,
    salesUserId: body.salesUserId ?? null,
    purchaseUserId: body.purchaseUserId ?? null,
    items
  }
}

function mapQuoteCreateFormToUpdate(form: Record<string, unknown>): Record<string, unknown> {
  const items = buildItemsFromQuoteCreateForm(form)
  return {
    mpn: form.mpn,
    remark: form.remark ?? null,
    salesUserId: form.salesUserId ?? null,
    purchaseUserId: form.purchaseUserId ?? null,
    items
  }
}

export const quoteApi = {
  /**
   * GET /api/v1/quotes（后端暂无筛选参数，在此做前端过滤）
   * 返回形状与旧 mock 一致：{ data, total }
   */
  async getList(params?: {
    keyword?: string
    status?: number
    rfqItemId?: string | null
  }) {
    const rows = await apiClient.get<unknown[]>('/api/v1/quotes')
    const list = Array.isArray(rows) ? rows : []
    let result = [...list] as Record<string, unknown>[]
    if (params?.keyword) {
      const kw = params.keyword.toLowerCase()
      result = result.filter((q) => {
        const items = (q.items ?? q.Items) as Record<string, unknown>[] | undefined
        const it0 = Array.isArray(items) && items.length > 0 ? items[0] : null
        const parts = [
          q.quoteCode,
          q.quoteNumber,
          q.rfqCode,
          q.mpn,
          q.customerName,
          q.salesUserName,
          it0?.brand,
          it0?.Brand,
          it0?.unitPrice,
          it0?.UnitPrice
        ]
          .filter((x) => x != null)
          .map((x) => String(x).toLowerCase())
        return parts.some((s) => s.includes(kw))
      })
    }
    if (params?.status !== undefined && params?.status !== null) {
      result = result.filter((q) => Number(q.status) === Number(params.status))
    }
    if (params?.rfqItemId != null && String(params.rfqItemId).trim() !== '') {
      const rid = String(params.rfqItemId).trim()
      result = result.filter((q) => String(q.rfqItemId ?? q.RfqItemId ?? '') === rid)
    }
    return { data: result, total: result.length }
  },

  async getById(id: string) {
    const data = await apiClient.get<unknown>(`/api/v1/quotes/${encodeURIComponent(id)}`)
    return { data }
  },

  async create(body: Record<string, unknown>) {
    const payload = mapToCreateQuoteRequest(body)
    const created = await apiClient.post<unknown>('/api/v1/quotes', payload)
    return { data: created }
  },

  async update(id: string, body: Record<string, unknown>) {
    const payload =
      body.quotePriceRows != null && Array.isArray(body.quotePriceRows)
        ? mapQuoteCreateFormToUpdate(body)
        : mapQuoteEditSimpleToUpdate(body)
    const updated = await apiClient.put<unknown>(`/api/v1/quotes/${encodeURIComponent(id)}`, payload)
    return { data: updated }
  },

  async delete(id: string) {
    await apiClient.delete(`/api/v1/quotes/${encodeURIComponent(id)}`)
  },

  async updateStatus(id: string, status: number) {
    await apiClient.patch(`/api/v1/quotes/${encodeURIComponent(id)}/status`, { status })
  }
}

export default quoteApi
